src/filemanager.cpp

```diff
@@ -0,0 +1,2301 @@
+}
+
+// ---------------------------------------------------------------------------
+
+bool DiskChunkCache::move_to_head(sqlite3_int64 chunk_id) {
+
+    sqlite3_int64 link_id = 0;
+    sqlite3_int64 prev = 0;
+    sqlite3_int64 next = 0;
+    sqlite3_int64 head = 0;
+    sqlite3_int64 tail = 0;
+    if (!get_links(chunk_id, link_id, prev, next, head, tail)) {
+        return false;
+    }
+
+    if (link_id == head) {
+        return true;
+    }
+
+    if (!update_links_of_prev_and_next_links(prev, next)) {
+        return false;
+    }
+
+    if (head) {
+        auto stmt = prepare("UPDATE linked_chunks SET prev = ? WHERE id = ?");
+        if (!stmt)
+            return false;
+        stmt->bindInt64(link_id);
+        stmt->bindInt64(head);
+        const auto ret = stmt->execute();
+        if (ret != SQLITE_DONE) {
+            pj_log(ctx_, PJ_LOG_ERROR, "%s", sqlite3_errmsg(hDB_));
+            return false;
+        }
+    }
+
+    return update_linked_chunks(link_id, 0, head) &&
+           update_linked_chunks_head_tail(link_id,
+                                          (link_id == tail) ? prev : tail);
+}
+
+// ---------------------------------------------------------------------------
+
+bool DiskChunkCache::move_to_tail(sqlite3_int64 chunk_id) {
+    sqlite3_int64 link_id = 0;
+    sqlite3_int64 prev = 0;
+    sqlite3_int64 next = 0;
+    sqlite3_int64 head = 0;
+    sqlite3_int64 tail = 0;
+    if (!get_links(chunk_id, link_id, prev, next, head, tail)) {
+        return false;
+    }
+
+    if (link_id == tail) {
+        return true;
+    }
+
+    if (!update_links_of_prev_and_next_links(prev, next)) {
+        return false;
+    }
+
+    if (tail) {
+        auto stmt = prepare("UPDATE linked_chunks SET next = ? WHERE id = ?");
+        if (!stmt)
+            return false;
+        stmt->bindInt64(link_id);
+        stmt->bindInt64(tail);
+        const auto ret = stmt->execute();
+        if (ret != SQLITE_DONE) {
+            pj_log(ctx_, PJ_LOG_ERROR, "%s", sqlite3_errmsg(hDB_));
+            return false;
+        }
+    }
+
+    return update_linked_chunks(link_id, tail, 0) &&
+           update_linked_chunks_head_tail((link_id == head) ? next : head,
+                                          link_id);
+}
+
+// ---------------------------------------------------------------------------
+
+void NetworkChunkCache::insert(PJ_CONTEXT *ctx, const std::string &url,
+                               unsigned long long chunkIdx,
+                               std::vector<unsigned char> &&data) {
+    auto dataPtr(std::make_shared<std::vector<unsigned char>>(std::move(data)));
+    cache_.insert(Key(url, chunkIdx), dataPtr);
+
+    auto diskCache = DiskChunkCache::open(ctx);
+    if (!diskCache)
+        return;
+    auto hDB = diskCache->handle();
+
+    // Always insert DOWNLOAD_CHUNK_SIZE bytes to avoid fragmentation
+    std::vector<unsigned char> blob(*dataPtr);
+    assert(blob.size() <= DOWNLOAD_CHUNK_SIZE);
```

Comment:
Should DOWNLOAD_CHUNK_SIZE be configurable?

Reply:
runtime configurable ? I don't see this as a urgent need. This is pretty much a development setting. We might reconsider that later i'd say